rllib/offline/offline_data.py

```diff
@@ -0,0 +1,152 @@
+from ray.rllib.utils.compression import unpack_if_needed
+from ray.rllib.utils.typing import EpisodeType
+
+logger = logging.getLogger(__name__)
+
+# TODO (simon): Implement schema mapping for users, i.e. user define
+# which row name to map to which default schema name below.
+SCHEMA = [
+    Columns.EPS_ID,
+    Columns.AGENT_ID,
+    Columns.MODULE_ID,
+    Columns.OBS,
+    Columns.ACTIONS,
+    Columns.REWARDS,
+    Columns.INFOS,
+    Columns.NEXT_OBS,
+    Columns.TERMINATEDS,
+    Columns.TRUNCATEDS,
+    Columns.T,
+    # TODO (simon): Add remove as soon as we are new stack only.
+    "agent_index",
+    "dones",
+    "unroll_id",
+]
+
+
+class OfflineData:
+    def __init__(self, config: Union[Dict[str, Any], AlgorithmConfig]):
+
+        self.config = config
+        self.is_multi_agent = (
+            config.is_multi_agent()
+            if isinstance(config, AlgorithmConfig)
+            else len(config.get("policies", {})) > 1
+        )
+        self.path = (
+            config.get("input_")
+            if isinstance(config.get("input_"), list)
+            else Path(config.get("input_"))
+        )
+        # Use `read_json` as default data read method.
+        self.data_read_method = config.get("data_read_method", "read_json")
+        self.compressed = config.get("compressed", False)
+        try:
+            self.data = getattr(ray.data, self.data_read_method)(self.path)
+            logger.info("Reading data from {}".format(self.path))
+            logger.info(self.data.schema())
+        except Exception as e:
+            logger.error(e)
+        self.batch_iterator = None
+
+    def sample(
+        self,
+        num_samples: int,
+        return_iterator: bool = False,
+        num_shards: int = 1,
+    ):
+        if (
+            not return_iterator
+            or return_iterator
+            and num_shards <= 1
+            and not self.batch_iterator
+        ):
+            self.batch_iterator = self.data.map_batches(
+                functools.partial(self._map_to_episodes, self.is_multi_agent)
+            ).iter_batches(
+                batch_size=num_samples,
+                prefetch_batches=1,
+                local_shuffle_buffer_size=num_samples * 10,
+            )
+
+        if return_iterator:
+            if num_shards > 1:
+                return self.data.map_batches(
+                    functools.partial(self._map_to_episodes, self.is_multi_agent)
+                ).streaming_split(n=num_shards, equal=True)
+            else:
+                return self.batch_iterator
+        else:
+            # Return a single batch
+            return next(iter(self.batch_iterator))["episodes"]
+
+    @staticmethod
+    def _map_to_episodes(
+        is_multi_agent: bool, batch: Dict[str, np.ndarray]
+    ) -> List[EpisodeType]:
+        """Maps a batch of data to episodes."""
+
+        episodes = []
+        # TODO (simon): Give users possibility to provide a custom schema.
```

Comment:
This is great!

I think we should support off-the-shelf in RLlib only few JSON schemas:
* SingleAgentEpisode
* MultiAgentEpisode
* SampleBatch (warn when used and point to new Episode format)
* NOT: MultiAgentBatch (too complicated)

Reply:
This is similar to my thoughts. Warn about old stack (batches) and build MA offline directly on top of episodes.